components/canvas/webgl_paint_task.rs

```diff
@@ -0,0 +1,119 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use canvas_msg::{CanvasWebGLMsg, CanvasCommonMsg, CanvasMsg};
+use compositing::windowing::{WindowMethods};
+use geom::size::Size2D;
+
+use gleam::gl;
+use gleam::gl::types::{GLint, GLsizei};
+
+use layers::platform::surface::{NativeGraphicsMetadata, NativePaintingGraphicsContext};
+use util::task::spawn_named;
+
+use skia::{SkiaGrGLNativeContextRef, SkiaSkNativeSharedGLContextCreate};
+use std::borrow::ToOwned;
+use std::sync::mpsc::{channel, Sender};
+use util::vec::byte_swap;
+
+pub struct WebGLPaintTask {
+    size: Size2D<i32>,
+}
+
+impl WebGLPaintTask {
+    fn new(size: Size2D<i32>, native_graphics_metadata: &NativeGraphicsMetadata) -> WebGLPaintTask {
+        let native_graphics_context = NativePaintingGraphicsContext::from_metadata(&native_graphics_metadata);
+        WebGLPaintTask::create(size, &native_graphics_context);
+        WebGLPaintTask {
+            size: size,
+        }
+    }
+
+    pub fn start(size: Size2D<i32>,
+                 native_metadata: NativeGraphicsMetadata) -> Sender<CanvasMsg> {
+        let (chan, port) = channel::<CanvasMsg>();
+        spawn_named("WebGLTask".to_owned(), move || {
+            let native_metadata = native_metadata;
+            let mut painter = WebGLPaintTask::new(size, &native_metadata);
+            painter.init();
+            loop {
+                match port.recv().unwrap() {
+                    CanvasMsg::WebGL(message) => {
+                        match message {
+                            CanvasWebGLMsg::Clear(mask) => painter.clear(mask),
+                            CanvasWebGLMsg::ClearColor(r, g, b, a) => painter.clear_color(r, g, b, a),
+                        }
+                    },
+                    CanvasMsg::Common(message) => {
+                        match message {
+                            CanvasCommonMsg::Close => break,
+                            CanvasCommonMsg::SendPixelContents(chan) => painter.send_pixel_contents(chan),
+                            CanvasCommonMsg::Recreate(size) => painter.recreate(size),
+                        }
+                    },
+                    CanvasMsg::Canvas2d(_) => panic!("Wrong message sent to WebGLTask"),
+                }
+            }
+        });
+        chan
+    }
+
+    fn create(size: Size2D<i32>,
+              native_graphics_context: &NativePaintingGraphicsContext) {
+        let graphics_context = native_graphics_context as *const _ as SkiaGrGLNativeContextRef;
+        // It creates an OpenGL context
```

Comment:
This is fine for now, but calling our Skia extension to do this is pretty ugly. Could you add a `FIXME` saying "provide a clean, cross-platform way to do this, e.g. in glutin"?